gordo_components/model/multi_regressor.py

```diff
@@ -0,0 +1,142 @@
+    >>> out = model.predict(X)
+    >>> assert out.shape == y.shape
+    """
+
+    models: List[BaseEstimator]
+
+    def __init__(
+        self,
+        base_estimator: Optional[BaseEstimator] = None,
+        n_jobs: int = -1,
+        verbose: bool = False,
+        pre_dispatch: Union[int, str] = "2*n_jobs",
+    ):
+        """
+
+        Parameters
+        ----------
+        base_estimator: BaseEstimator
+            Model which is used to predict individual targets of y.
+            Internally, this model is cloned and used for each axis=1 of y.
+        n_jobs : int or None, optional (default=None)
+            The number of CPUs to use to do the computation.
+            ``None`` means 1 unless in a :obj:`joblib.parallel_backend` context.
+            ``-1`` means using all processors. See :term:`Glossary <n_jobs>`
+            for more details.
+        verbose : integer, optional
+            The verbosity level.
+        pre_dispatch : int, or string, optional
+            Controls the number of jobs that get dispatched during parallel
+            execution. Reducing this number can be useful to avoid an
+            explosion of memory consumption when more jobs get dispatched
+            than CPUs can process. This parameter can be:
+                - None, in which case all the jobs are immediately
+                  created and spawned. Use this for lightweight and
+                  fast-running jobs, to avoid delays due to on-demand
+                  spawning of the jobs
+                - An int, giving the exact number of total jobs that are
+                  spawned
+                - A string, giving an expression as a function of n_jobs,
+                  as in '2*n_jobs'
+        """
+        self.models = []
+        self.base_estimator = (
+            base_estimator
+            if base_estimator is not None
+            else RandomForestRegressor(n_estimators=10)
+        )
+        self.pre_dispatch = pre_dispatch
+        self.n_jobs = n_jobs
+        self.verbose = verbose
+
+    def _clone_base_estimator(self) -> BaseEstimator:
+        """
+        Clone, on demand, the original base_estimator passed into the init
+        """
+        return clone(self.base_estimator)
+
+    @staticmethod
+    def _fit_model(model: BaseEstimator, X: np.ndarray, y: np.ndarray) -> BaseEstimator:
+        """Fit a single model on X and y"""
+        model.fit(X, y)
+        return model
+
+    def fit(self, X: np.ndarray, y: np.ndarray) -> "MultiTargetRegressor":
+        """
+        Fit the MultiTargetRegressor on X and y, where y is assumed, but not required
+        to have multiple dimensions on axis=1. Internally a ``base_estimator`` will be used
+        with X and the specific index of y. This is done in parallel.
+
+        Parameters
+        ----------
+        X: np.ndarray
+        y: np.ndarray
+
+        Returns
+        -------
+        self
+        """
+        parallel = Parallel(
```

Comment:
If I understand correctly this will use the `loky` backend, which causes new processes. There is some fancyness to how the shared data (X) is shared among them, it writes it to `/dev/shm` (a RAM backed FS) and then the different processes reads from there.  But in docker this seems to be default 64MB. Luckily joblib [now](https://github.com/joblib/joblib/pull/609) detects the size, and does not use it if it is too smal. So I think when we run this under docker joblib would "transparantly" copy X into each of the subprocesses, causing more memory usage. One way to fix this is to set `require='sharedmem'` to `Parallel`. Then it will use threads, but I dont know how much of a problem that is. If the random forrest uses numpy then the GIL is released whenever numpy does something. But anyway probably better to be a bit slower than to crash. But feel free to make a parameter out of it, so you can choose the "use-much-memory-but-maybe-be-faster" option if you want.

Reply:
:boom: :computer: :boom: